doc: clarify -timeout and -peertimeout config options

glozow opened the pull request:
The debug-only option `-peertimeout` is used to delay `InactivityCheck()`, whereas the `-timeout` option specifies socket timeouts (`nConnectTimeout`). The current descriptions are a bit misleading and hard to tell apart. I think it would save dev/review time to update them 🤷

glozow committed:
[doc] clarify -peertimeout and -timeout descriptions

RandyMcMillan commented:
@glozow 

This PR has a broader scope than docs - maybe add Utils/Log/Libs label or cli ?

zoltan-konder commented on src/init.cpp:
```diff
@@ -455,7 +455,7 @@ void SetupServerArgs(NodeContext& node)
     argsman.AddArg("-seednode=<ip>", "Connect to a node to retrieve peer addresses, and disconnect. This option can be specified multiple times to connect to multiple nodes.", ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
     argsman.AddArg("-networkactive", "Enable all P2P network activity (default: 1). Can be changed by the setnetworkactive RPC command", ArgsManager::ALLOW_BOOL, OptionsCategory::CONNECTION);
     argsman.AddArg("-timeout=<n>", strprintf("Specify connection timeout in milliseconds (minimum: 1, default: %d)", DEFAULT_CONNECT_TIMEOUT), ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
-    argsman.AddArg("-peertimeout=<n>", strprintf("Specify p2p connection timeout in seconds. This option determines the amount of time a peer may be inactive before the connection to it is dropped. (minimum: 1, default: %d)", DEFAULT_PEER_CONNECT_TIMEOUT), ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::CONNECTION);
```
I don't like how you just removed p2p connection. It was good how it was.

zoltan-konder replied:
There was no issue with any of the text. It was written very clearly: 

`Specify p2p connection timeout in seconds. This option determines the amount of time a peer may be inactive before the connection to it is dropped`

zoltan-konder commented on src/init.cpp:
```diff
@@ -455,7 +455,7 @@ void SetupServerArgs(NodeContext& node)
     argsman.AddArg("-networkactive", "Enable all P2P network activity (default: 1). Can be changed by the setnetworkactive RPC command", ArgsManager::ALLOW_BOOL, OptionsCategory::CONNECTION);
     argsman.AddArg("-timeout=<n>", strprintf("Specify connection timeout in milliseconds (minimum: 1, default: %d)", DEFAULT_CONNECT_TIMEOUT), ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
-    argsman.AddArg("-peertimeout=<n>", strprintf("Specify p2p connection timeout in seconds. This option determines the amount of time a peer may be inactive before the connection to it is dropped. (minimum: 1, default: %d)", DEFAULT_PEER_CONNECT_TIMEOUT), ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::CONNECTION);
+    argsman.AddArg("-peertimeout=<n>", strprintf("Specify a timeout delay in seconds. After connecting to a peer, wait this amount of time before deciding to disconnect based on inactivity. (minimum: 1, default: %d)", DEFAULT_PEER_CONNECT_TIMEOUT), ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::CONNECTION);
```
```suggestion
    argsman.AddArg("-peertimeout=<n>", strprintf("Specify a timeout delay in seconds. After connecting to a peer, wait this amount of time before disconnecting based on inactivity. (minimum: 1, default: %d)", DEFAULT_PEER_CONNECT_TIMEOUT), ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::CONNECTION);
```

zoltan-konder replied:
No matter how I look it, I don't like your version of the text @glozow.

Even `based on inactivity` should be changed to `in the case of inactivity`

`Specify a connection timeout delay in seconds. After connecting to a peer, wait this amount of time before disconnecting in the case of inactivity.` What connection? P2P...

zoltan-konder reviewed (changes requested):
**deciding** to disconnect sound bad, who decides? To make a decision you need a human being. A program can't "deciding" what to do.

zoltan-konder commented:
@laanwj I kindly ask you not to merge this PR until she don't fix it. This PR is not worth to be merged in this form how she submitted it. Does more damage than good.

laanwj commented:
@zoltan-konder There's no need to tag me personally here, Also to be honest I think it's kind of bothersome how you behave. This is a simple documentation change and while you're free to express your disagreement, the way you do so is unacceptable.

zoltan-konder commented:
Not like I care what is acceptable and what is not for you personally. This is the Bitcoin repository and someone removing things like "p2p connection" and other text because she don't understand what it means. 

I'm kind a had enough from self declared "Bitcoin experts" etc who sending this kind of PR to this repository. I'm here to voice my "disagreements" of course but she is so egoist that not writing anything. 

Fix the PR because the PR submitted by @glozow unacceptable not me!

jnewbery commented:
ACK f2306cbb9f564815b015fdbdb80410299fd22dee although I agree with @ghost that "P2P" could be added back.

> There was no issue with any of the text. It was written very clearly

@ghost: I disagree. The text as currently written states that if an p2p connection stops responding to messages for -peertimeout seconds, then we'll disconnect. This isn't what the config option does. It controls how long we'll wait after we connect before we start running inactivity checks. Those inactivity checks always use the hardcoded TIMEOUT_INTERVAL, which is 20 minutes.

hebasto commented on src/init.cpp:
```diff
@@ -455,7 +455,7 @@ void SetupServerArgs(NodeContext& node)
     argsman.AddArg("-networkactive", "Enable all P2P network activity (default: 1). Can be changed by the setnetworkactive RPC command", ArgsManager::ALLOW_BOOL, OptionsCategory::CONNECTION);
     argsman.AddArg("-timeout=<n>", strprintf("Specify connection timeout in milliseconds (minimum: 1, default: %d)", DEFAULT_CONNECT_TIMEOUT), ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
-    argsman.AddArg("-peertimeout=<n>", strprintf("Specify p2p connection timeout in seconds. This option determines the amount of time a peer may be inactive before the connection to it is dropped. (minimum: 1, default: %d)", DEFAULT_PEER_CONNECT_TIMEOUT), ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::CONNECTION);
+    argsman.AddArg("-peertimeout=<n>", strprintf("Specify a timeout delay in seconds. After connecting to a peer, wait this amount of time before deciding to disconnect based on inactivity. (minimum: 1, default: %d)", DEFAULT_PEER_CONNECT_TIMEOUT), ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::CONNECTION);
```
nit: maybe drop the last full stop as done for most of the help strings?
```suggestion
    argsman.AddArg("-peertimeout=<n>", strprintf("Specify a timeout delay in seconds. After connecting to a peer, wait this amount of time before deciding to disconnect based on inactivity (minimum: 1, default: %d)", DEFAULT_PEER_CONNECT_TIMEOUT), ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::CONNECTION);
```

hebasto reviewed (approved):
ACK f2306cbb9f564815b015fdbdb80410299fd22dee

I see this PR as a kind of follow up of #20927.

Currently, on master:
```
$ src/bitcoind -help-debug | grep -A 2 -e '-timeout'
  -timeout=<n>
       Specify connection timeout in milliseconds (minimum: 1, default: 5000)

$ src/bitcoind -help-debug | grep -A 4 -e '-peertimeout'
  -peertimeout=<n>
       Specify p2p connection timeout in seconds. This option determines the
       amount of time a peer may be inactive before the connection to it
       is dropped. (minimum: 1, default: 60)

```

With this PR (f2306cbb9f564815b015fdbdb80410299fd22dee):
```
$ src/bitcoind -help-debug | grep -A 4 -e '-peertimeout'
  -peertimeout=<n>
       Specify a timeout delay in seconds. After connecting to a peer, wait
       this amount of time before deciding to disconnect based on
       inactivity. (minimum: 1, default: 60)

```

that describes the logic of `CConnman::InactivityCheck` much better:
https://github.com/bitcoin/bitcoin/blob/bf100f8170770544fb39ae6802175c564cde532f/src/net.cpp#L1225-L1229

jnewbery commented:
> Currently, on master:
> ```
> $ src/bitcoind -help-debug | grep -A 2 -e '-timeout'
>   -timeout=<n>
>        Specify connection timeout in milliseconds (minimum: 1, default: 5000)```

This config option is _also_ under-documented! The timeout this is referring to is the socket timeout on initial connection (i.e. if we poll or select don't return a file descriptor because the initial connection timed out)

MarcoFalke commented:
Also, `nConnectTimeout` is a global (bad!!) and could be a const member of connman instead

jnewbery commented:
> Also, nConnectTimeout is a global (bad!!) and could be a const member of connman instead

I'm not sure if connman is the right home for this. It's a low-level socket configuration, and could live entirely in netbase. It seems strange that connman takes a global from netbase.cpp and then passes it to netbase through `ConnectThroughProxy()` or `ConnectSocketDirectly()`.

In any case, I think we've wondered off topic a little bit :)

glozow commented:
Thanks @hebasto and @jnewbery for your kind and thoughtful feedback ☺️ 
I've addressed your comments and edited the `-timeout` help string a little bit so that they're more distinguishable.

MarcoFalke commented:
ACK eecb7ab105a4a59d09cd55b124c5ad563846fe11 nice doc fixup

jnewbery commented:
ACK eecb7ab105a4a59d09cd55b124c5ad563846fe11
